enterprise/backend/src/metabase_enterprise/task/cache.clj

```diff
@@ -4,23 +4,237 @@
+                         ;; This is a safety check so that we don't scan all of query_execution -- if a query
+                         ;; has not been excuted at all in the last month (including cache hits) we won't bother
+                         ;; refreshing it again.
+                         [:>= :qe.started_at (duration-ago {:duration 30 :unit "days"})]
+                         [:= :qe.error nil]
+                         [:= :qe.is_sandboxed false]
+                         (if parameterized?
+                           [:and
+                            [:= :qe.parameterized true]
+                             ;; Only rerun a parameterized query if it's had a cache hit within the last caching window
+                            [:= :qe.cache_hit true]
+                             ;; Don't factor the last cache refresh into whether we should rerun a parameterized query
+                            [:not= :qe.context (name :cache-refresh)]]
+                           [:= :qe.parameterized false])]
+              :group-by [:q.query_hash :q.query :qe.card_id]}}))]
+    {:select [:u.query :u.card-id :u.count]
+     :from   [[{:union queries} :u]]}))
+
+(defn- cache-configs->card-ids
+  "Takes a list of cache configs and returns a map from cache configs to relevant card IDs. For cache configs defined on
+  a dashboard, this includes all cards on the dashboard."
+  [cache-configs]
+  (let [dashboard-ids          (->> (filter #(= (:model %) "dashboard") cache-configs)
+                                    (map :model_id))
+        dashboards             (when (seq dashboard-ids)
+                                 (-> (t2/select :model/Dashboard {:where [:in :id dashboard-ids]})
+                                     (t2/hydrate :dashcards)))
+        dashboard-id->card-ids (into {}
+                                     (map (fn [dashboard] [(:id dashboard) (mapv :card_id (:dashcards dashboard))])
+                                          dashboards))
+        config-to-card-ids     (map (fn [config]
+                                      (let [model (:model config)
+                                            model-id (:model_id config)]
+                                        [config
+                                         (cond
+                                           (= model "dashboard") (get dashboard-id->card-ids model-id [])
+                                           (= model "question") [model-id]
+                                           :else [])]))
+                                    cache-configs)]
+    (into {} config-to-card-ids)))
+
+(defn- select-parameterized-queries
+  [parameterized-queries]
+  (apply concat
+         (-> (group-by :card-id parameterized-queries)
+             (update-vals
+              (fn [queries]
+                (->> queries
+                     (sort-by :count >)
+                     (take *parameterized-queries-to-rerun-per-card*))))
```

Comment:
nit: I think this could be a `map`, then `group-by` instead, and that would be a little clearer

Reply:
Oh, no it isn't since you take _n_ for each card-id. Thanks for pointing that out!

Reply:
I'll add a comment here to help explain the logic